fairscale/experimental/wgit/signal_sparsity.py

```diff
@@ -112,6 +155,70 @@ def sst_dst_to_dense(self, sst: Tensor, dst: Tensor = None) -> Tensor:
         """
         pass
 
+    def _transform(self, dense: Tensor) -> Tensor:
+        """
+        Applies the transformation algorithm chosen by the caller to the input dense tensor.
+        Returns the transformed tensor.
+
+        Args:
+            dense (Tensor):
+                Input dense tensor.
+        """
+        if self._algo is Algo.FFT:
+            transform = torch.fft.fft  # type: ignore
+        elif self._algo is Algo.DCT:
+            raise NotImplementedError
+        return transform(dense)
+
+    def _sparse_mask(
+        self,
+        in_tensor: Tensor,
+        top_k_element: Optional[int] = None,
+        top_k_percent: Optional[float] = None,
+        dim: int = -1,
+    ) -> Tensor:
+        """Returns a mask for a tensor with a certain sparsity level corresponding to either
+        the top_k_percent or the top_k_element of the tensor.
+
+        Args:
+            in_tensor (Tensor)
+                input torch tensor for which sparse mask is generated.
+            top_k_element (float, optional)
+                Number of top-k elements to retain in the dense tensor corresponding to the mask.
+            top_k_percent (float, optional)
+                Percent of top-k elements to retain in the dense tensor corresponding to the mask.
+            dim (int, optional):
+                The dimension on which the top-k is done..
+                E.g. -1 is the last dim. None means flatten and top-k on all dims.
+                There is no way to specify multiple dims other than None.
+        """
+        abs_tensor = torch.abs(in_tensor)
+
+        # when the criterion is top_k_percent
+        if self._dst_top_k_percent is not None or self._sst_top_k_percent is not None:
+            sparsity = 1 - top_k_percent  # target sparsity in the unit interval [0,1]
+            if sparsity == 0.0:  # if sparsity is zero, we want a mask with all 1's
+                threshold = torch.tensor(float("-Inf"), device=in_tensor.device).unsqueeze(dim)
+            else:
+                threshold = torch.quantile(abs_tensor, sparsity, dim).unsqueeze(dim)  # type: ignore
+            sps_mask = in_tensor >= threshold
+
+        # when the criterion is top_k_elements
+        elif self._dst_top_k_element is not None or self._sst_top_k_element is not None:
```

Comment:
Let's chat in VC. I don't know why you need 2 cases and a threshold above. I was thinking we can reduce the percent into an element count and always do a top-K. I actually don't know why a mask is needed. Maybe you are trying to implement value based percentage, not count based percentage?